designs/limits.md

```diff
@@ -0,0 +1,60 @@
+
+
+## Current State
+
+To address the runaway-scaling problem the current fix in place is to detect if the kubelet for a worker node has never reported its status to the K8s control plane. If it's been longer than 15 minutes, Karpenter assumes that there's a hard failure mode due to which this worker node will never become healthy and terminates the worker node. If the condition map of the node object in the API Server says `NodeStatusNeverUpdated` then we use that as an indicator of the node having never come up.
+
+This fix ensures that if there are other scenarios where a worker node has become unhealthy due to a network partition or power outage in a availability zone, we don't terminate those worker nodes. It's important we don't make the static stability of a cluster worse during such an event. On the other hand, if there is an edge case where worker nodes come online and soon go offline, it will lead to runaway scaling again. This edge case should be unlikely to happen in the near term, so this document focuses on just the ability to limit costs within Karpenter. That way even if runaway scaling does occur there's a way to bound it. A longer-term solution to handle the runaway problem will be discussed separately.
+
+
+## Proposed Solution for Limits
+
+There are two broad forms of limiting we could apply. The first is that we could introduce a limit to the number of in-flight worker node being provisioned at a point in time. A worker node that's in the `NotReady` state could be considered to be in-flight. The second form is an absolute limit of the number of resources Karpenter can provision.
+
+
+### **In-flight limit**
+
+```yaml
+spec:
+  limits:
+    unready: 20% # StringOrInt
+```
+
+The good bit about this approach is that we don't artificially constrain how many total worker nodes can be spun up by Karpenter, while also making sure that if we're launching worker nodes that aren't healthy we stop the scaling and save costs.
+
+The two main problems with this approach though are -
+1. This limit while meant to constrain the number of unhealthy worker nodes in a cluster, will also inihibit the rate at which Karpenter can respond to pods that aren't schedulable. This somewhat goes against the goal of minimizing launch times of workers.
+2. While this helps ensure that costs don't increase due to runaway scaling, it won't help those who want a stricter cap on the amount of resources that's being provisioned even when nodes are otherwise healthy.
+
+### **Absolute limit**
+
+This is a much simpler semantic to follow, where you provide an upper cap on the amount of resources being provisioned irrespective of their health. This is more in-line with alternative solutions like CA and should be easier to reason with.
+
+The actual limit of the number of resources being spun up can be defined either through something similar to resource requests or just a flat count of instances.
```

Comment:
> Should we consider supporting number of pods as an upper limit?

I'm not sure how I'd be able to set that value reasonably. In a cluster I'd probably have multiple deployments running with HPA behind the scenes so it'll be difficult for me to keep this number up to date.  I also think since Karpenter provisions your infrastructure, it's limits being modeled as infra characteristics like cpu / mem / gpu would be a little more intuitive too. 

> Also, if we're going to do absolute limits should we break it out by capacity type?

Ooh this would be nice. I know we just made capacity-type a well known concept too so I wouldn't be against it. Although I suppose we could be even more generic where you could also break it down by other well known constraints like availability-zone? With AZs, we could kind of enforce Karpenter does some form of balancing across zones for higher availability. I wonder if this is too much complexity to start with?

Reply:
I think the balancing is probably a bit advanced and maybe overloads this limits concept a bit, but I think factoring in having the limit be per capacity type or across all capacity types is still reasonable. Especially since a big part of why this limit will exist is to control costs and different capacity types can (and in AWS's case certainly do) have drastic cost implications.

Reply:
Yeah that's fair on the balancing bit. Following up on the conversation in the working group meeting though, this might be a little difficult to represent and use if we keep adding new dimensions to split the limits. Even with say capacity type, it might end up looking something like - 

```
  limits:
    resources:
        cpu: 1000 # irrespective of capacity type?
      capacity-type: on-demand
        cpu: 1000
      capacity-type: spot
        cpu: 500

or 

  limits:
    resources:
        cpu:
          capacity-types: 
             spot: 500
             on-demand: 1000
         mem:
  ....
```

And it'd start to get more complicated from there. I think to the other feedback that these limits are more to bound any scaling when something seems really off and isn't so much about precise cost control, I think we could probably leave this optimization out. Does that sound reasonable to you?